middle_end/flambda2/types/meet_and_join_new.ml

```diff
@@ -780,79 +786,46 @@ and meet_head_of_kind_naked_immediate env (t1 : TG.head_of_kind_naked_immediate)
     (t2 : TG.head_of_kind_naked_immediate) :
     TG.head_of_kind_naked_immediate meet_result =
   let module I = Targetint_31_63 in
-  let keep_side side : _ meet_result =
-    match side with
-    | Left -> Ok (Left_input, env)
-    | Right -> Ok (Right_input, env)
-  in
-  let keep_other_side side : _ meet_result =
-    match side with
-    | Left -> Ok (Right_input, env)
-    | Right -> Ok (Left_input, env)
+  (* First compute the result type *)
+  let meet_immediates =
+    meet_unknown
+      (set_meet (module I.Set) ~of_set:Fun.id)
+      ~contents_is_bottom:I.Set.is_empty
   in
-  let meet_with_shape ~rebuild ty shape side =
-    map_result ~f:rebuild
-      (match side with Left -> meet env ty shape | Right -> meet env shape ty)
+  let meet_relations = reverse_set_meet (module TG.RelationSet) in
+  let result =
+    combine_results2 env ~meet_a:meet_immediates ~meet_b:meet_relations
+      ~left_a:t1.immediates ~right_a:t2.immediates ~left_b:t1.relations
+      ~right_b:t2.relations ~rebuild:(fun immediates relations ->
+        TG.Head_of_kind_naked_immediate.create ~immediates ~relations)
   in
-  let is_int_immediate ~is_int_ty ~immediates ~is_int_side =
-    let rebuild = TG.Head_of_kind_naked_immediate.create_is_int in
-    match I.Set.mem I.zero immediates, I.Set.mem I.one immediates with
-    | false, false -> Bottom
-    | true, true -> keep_side is_int_side
-    | true, false ->
-      meet_with_shape ~rebuild is_int_ty MTC.any_block is_int_side
-    | false, true ->
-      meet_with_shape ~rebuild is_int_ty MTC.any_tagged_immediate is_int_side
-  in
-  let get_tag_immediate ~get_tag_ty ~immediates ~get_tag_side =
-    if I.Set.is_empty immediates
-    then keep_other_side get_tag_side
-    else
-      let tags =
-        I.Set.fold
-          (fun tag tags ->
-            match Tag.create_from_targetint tag with
-            | Some tag -> Tag.Set.add tag tags
-            | None -> tags (* No blocks exist with this tag *))
-          immediates Tag.Set.empty
+  (* Now, use the relations to do some reduction *)
+  match result with
+  | Bottom -> result
+  | Ok (new_value, _env) -> (
+    (* Note: the env is not discarded, [result] itself is returned directly or
+       passed to the fold function *)
+    match extract_value new_value t1 t2 with
+    | { immediates = Unknown; relations = _ } ->
+      (* No reduction to do *)
+      result
+    | { immediates = Known is; relations } ->
+      (* Update the result by reduction, following the relations *)
+      let reduce relation (result : _ meet_result) : _ meet_result =
+        match result with
+        | Bottom -> Bottom
```

Comment:
Here it is `match result with Bottom -> Bottom` and higher, it is `match result with Bottom -> result`
It should be coherent I think. I tend to prefer `Bottom -> Bottom`